sdks/python/apache_beam/transforms/ptransform.py

```diff
@@ -97,29 +100,112 @@ def visit(self, node, replacements):
+def _allocate_materialized_pipeline(pipeline):
+  with _pipeline_materialization_lock:
+    pipeline_id = id(pipeline)
+    _pipeline_materialization_cache[pipeline_id] = {}
+
+
+def _allocate_materialized_result(pipeline):
+  with _pipeline_materialization_lock:
+    pipeline_id = id(pipeline)
+    if pipeline_id not in _pipeline_materialization_cache:
+      raise ValueError('Materialized pipeline is not allocated for result '
+                       'cache.')
+    result_id = len(_pipeline_materialization_cache[pipeline_id])
+    result = _MaterializedResult(pipeline_id, result_id)
+    _pipeline_materialization_cache[pipeline_id][result_id] = result
+    return result
+
+
+def _get_materialized_result(pipeline_id, result_id):
+  with _pipeline_materialization_lock:
+    if pipeline_id not in _pipeline_materialization_cache:
+      raise Exception(
+          'Materialization in out-of-process and remote runners is not yet '
+          'supported.')
+    return _pipeline_materialization_cache[pipeline_id][result_id]
+
+
+def _release_materialized_pipeline(pipeline):
+  with _pipeline_materialization_lock:
+    pipeline_id = id(pipeline)
+    del _pipeline_materialization_cache[pipeline_id]
+
+
+class _MaterializedResult(object):
+  def __init__(self, pipeline_id, result_id):
+    self._pipeline_id = pipeline_id
+    self._result_id = result_id
+    self.elements = []
+
+  def __reduce__(self):
+    # When unpickled (during Runner API roundtrip serailization), get the
+    # _MaterializedResult object from the cache so that values are written
+    # to the original _MaterializedResult when run in eager mode.
+    return (_get_materialized_result, (self._pipeline_id, self._result_id))
+
+
 class _MaterializedDoOutputsTuple(pvalue.DoOutputsTuple):
-  def __init__(self, deferred, pvalue_cache):
+  def __init__(self, deferred, results_by_tag):
     super(_MaterializedDoOutputsTuple, self).__init__(
         None, None, deferred._tags, deferred._main_tag)
     self._deferred = deferred
-    self._pvalue_cache = pvalue_cache
+    self._results_by_tag = results_by_tag
 
   def __getitem__(self, tag):
-    # Simply accessing the value should not use it up.
-    return self._pvalue_cache.get_unwindowed_pvalue(
-        self._deferred[tag], decref=False)
+    if tag not in self._results_by_tag:
+      raise KeyError(
+          'Tag %r is not a a defined output tag of %s.' % (
+              tag, self._deferred))
+    return self._results_by_tag[tag].elements
+
+
+class _AddMaterializationTransforms(_PValueishTransform):
 
+  def _materialize_transform(self, pipeline):
+    result = _allocate_materialized_result(pipeline)
 
-class _MaterializePValues(_PValueishTransform):
-  def __init__(self, pvalue_cache):
-    self._pvalue_cache = pvalue_cache
+    # Need to define _MaterializeValuesTransform here to avoid circular
+    # dependencies.
+    from apache_beam import DoFn
+    from apache_beam import ParDo
+
+    class _MaterializeValuesTransform(DoFn):
```

Comment:
s/_MaterializeValuesTransform/_MaterializeValuesDoFn/g

If we actually introduced a transform here, it could be easier for runners to swap it out for their own (possibly remote supporting) one in the future, but that's probably future work.